hail/src/main/scala/is/hail/expr/ir/ForwardLets.scala

```diff
@@ -27,45 +27,56 @@ object ForwardLets {
               !ContainsAgg(value)) &&
             !ContainsAggIntermediate(value)
         )
-      }
 
-      ir match {
-        case l: Block =>
-          val keep = new BoxedArrayBuilder[Binding]
-          val refs = uses(l)
-          val newEnv = l.bindings.foldLeft(env) {
-            case (env, Binding(name, value, scope)) =>
-              val rewriteValue = rewrite(value, env.promoteScope(scope)).asInstanceOf[IR]
-              if (
-                rewriteValue.typ != TVoid
-                && shouldForward(rewriteValue, refs.filter(_.t.name == name), l, scope)
-              ) {
-                env.bindInScope(name, rewriteValue, scope)
-              } else {
-                keep += Binding(name, rewriteValue, scope)
-                env
+      def rewrite(ir: BaseIR, env: BindingEnv[IR]): BaseIR =
+        ir match {
+          case l: Block =>
+            val keep = new BoxedArrayBuilder[Binding]
+            val refs = uses(l)
+            val newEnv = l.bindings.foldLeft(env) {
+              case (env, Binding(name, value, scope)) =>
+                val rewriteValue = rewrite(value, env.promoteScope(scope)).asInstanceOf[IR]
+                if (
+                  rewriteValue.typ != TVoid
+                  && shouldForward(rewriteValue, refs.filter(_.t.name == name), l, scope)
+                ) {
+                  env.bindInScope(name, rewriteValue, scope)
+                } else {
+                  keep += Binding(name, rewriteValue, scope)
+                  env
+                }
+            }
+
+            val newBody = rewrite(l.body, newEnv).asInstanceOf[IR]
+            if (keep.isEmpty) newBody
+            else Block(keep.result(), newBody)
+
+          case x @ Ref(name, _) =>
+            env.eval
+              .lookupOption(name)
+              .map { forwarded =>
+                if (uses.lookup(defs.lookup(x)).count(_.t.name == name) > 1) forwarded.deepCopy()
+                else forwarded
               }
-          }
+              .getOrElse(x)
+          case _ =>
+            ir.mapChildrenWithIndex((ir1, i) =>
+              rewrite(ir1, env.extend(Bindings.get(ir, i).dropBindings))
+            )
+        }
 
-          val newBody = rewrite(l.body, newEnv).asInstanceOf[IR]
-          if (keep.isEmpty) newBody
-          else Block(keep.result(), newBody)
+      val ir = rewrite(ir1, BindingEnv(Env.empty, Some(Env.empty), Some(Env.empty)))
 
-        case x @ Ref(name, _) =>
-          env.eval
-            .lookupOption(name)
-            .map { forwarded =>
-              if (uses.lookup(defs.lookup(x)).count(_.t.name == name) > 1) forwarded.deepCopy()
-              else forwarded
-            }
-            .getOrElse(x)
-        case _ =>
-          ir.mapChildrenWithIndex((ir1, i) =>
-            rewrite(ir1, env.extend(Bindings.get(ir, i).dropBindings))
+      try
+        TypeCheck(ctx, ir)
+      catch {
```

Comment:
This is the major change in this class - move `TypeCheck` here rather than call in `Optimize`